src/meta/src/manager/event_log.rs

```diff
@@ -0,0 +1,297 @@
+                        }
+                    };
+                },
+            }
+        }
+    };
+    let worker_join_handle = tokio::spawn(worker_loop);
+    EventLogManger::new(
+        meta_store,
+        event_tx,
+        (worker_join_handle, shutdown_tx),
+        enabled,
+    )
+}
+
+/// Gets a lease to attach it to new keys. The function
+/// - either returns the old lease,
+/// - or grant a new lease, if the old lease is likely to expire within `lease_ttl_sec`.
+///
+/// The idea is to make a trade-off between total number of lease granted and timeliness of key expiration.
+/// New lease's ttl is set to `lease_ttl_sec` + `lease_min_interval_sec`.
+/// Whenever a lease's remaining ttl is less than `lease_ttl_sec`, it should not be used for new keys anymore.
+async fn may_get_new_lease(
+    etcd_client: &EtcdRetryClient,
+    old_lease: Option<LeaseInfo>,
+    lease_ttl_sec: u64,
+) -> Result<Option<LeaseInfo>, etcd_client::Error> {
+    // `lease_min_interval_sec` is to avoid granting a new lease for each put op.
+    // When a lease has lived for over `lease_min_interval_sec`, a new lease should be granted to meet `lease_ttl_sec`.
+    let lease_min_interval_sec = std::cmp::min(3600, lease_ttl_sec);
+    let get_new = match &old_lease {
+        None => true,
+        Some(l) => l.1.elapsed().as_secs() >= lease_min_interval_sec,
+    };
+    if !get_new {
+        return Ok(old_lease);
+    }
+    let new_lease_id = etcd_client
+        .grant((lease_ttl_sec + lease_min_interval_sec) as i64, None)
+        .await?
+        .id();
+    Ok(Some((new_lease_id, Instant::now())))
+}
+
+pub type EventLogMangerRef = Arc<EventLogManger>;
+type ShutdownSender = tokio::sync::oneshot::Sender<()>;
+
+pub struct EventLogManger {
+    meta_store: MetaStoreRef,
+    event_tx: EventLogSender,
+    worker_join_handle: RwLock<Option<(JoinHandle<()>, ShutdownSender)>>,
+    enabled: bool,
+}
+
+impl EventLogManger {
+    pub fn new(
+        meta_store: MetaStoreRef,
+        event_tx: EventLogSender,
+        worker_join_handle: (JoinHandle<()>, ShutdownSender),
+        enabled: bool,
+    ) -> Self {
+        if !enabled {
+            tracing::info!("event log is disabled");
+        }
+        Self {
+            meta_store,
+            event_tx,
+            worker_join_handle: RwLock::new(Some(worker_join_handle)),
+            enabled,
+        }
+    }
+
+    #[cfg(any(test, feature = "test"))]
+    pub fn for_test(meta_store: MetaStoreRef) -> Self {
+        let (event_tx, _event_rx) = tokio::sync::mpsc::channel(BUFFER_SIZE);
+        Self {
+            meta_store,
+            event_tx,
+            worker_join_handle: Default::default(),
+            enabled: false,
+        }
+    }
+
+    pub fn take_join_handle(&self) -> Option<(JoinHandle<()>, ShutdownSender)> {
+        self.worker_join_handle.write().take()
+    }
+
+    pub fn add_event_logs(&self, event_logs: Vec<EventLog>) -> MetaResult<()> {
```

Comment:
Nit: Can we `event_logs: impl IntoIterator<EventLog>` here? Then `.add_event_logs(vec![event_log])` can be `.add_event_logs([event_log])`